### PR TITLE
OPTION_DO_WORK_FREQUENCY_IN_MS

### PR DESCRIPTION
**Pull request by YoDaMa**
## Code Changes
- Add OPTION_DO_WORK_FREQUENCY_IN_MS to sample and uncomment the line for logging in the convenience sample.
- Improve sample inline comments for convenience sample.
- Update requirement docs for OPTION_DO_WORK_FREQUENCY_IN_MS
- Add unit testing for OPTION_DO_WORK_FREQUENCY_IN_MS
- Change do_work_freq_ms from type uint16_t to type tickcounter_ms_t for clarity.

### CODE REVIEW

File: iothub_client/samples/iothub_convenience_sample/iothub_convenience_sample.c [hunk 1]
```diff
@@ -240,9 +241,14 @@ int main(void)
 
         // Set any option that are neccessary.
         // For available options please see the iothub_sdk_options.md documentation
+#ifndef SAMPLE_HTTP
+        bool traceOn = true;
+        (void)IoTHubDeviceClient_SetOption(device_handle, OPTION_LOG_TRACE, &traceOn);
+#endif
+        // ms_delay can be between 1 and 100 milliseconds
```
Comment (jspaith) [inline iothub_client/samples/iothub_convenience_sample/iothub_convenience_sample.c]: Please describe what this does or link to docs that do.

Comment (bikamani) [reply iothub_client/samples/iothub_convenience_sample/iothub_convenience_sample.c]: May be add a comment for this ifndef to make it more clear.
// Can not set this options in HTTP

File: iothub_client/samples/iothub_convenience_sample/iothub_convenience_sample.c [hunk 2]
```diff
@@ -240,9 +241,14 @@ int main(void)
...
+        tickcounter_ms_t ms_delay = 10;
+        (void)IoTHubDeviceClient_SetOption(device_handle, OPTION_DO_WORK_FREQUENCY_IN_MS, &ms_delay);
```
Comment (bikamani) [inline iothub_client/samples/iothub_convenience_sample/iothub_convenience_sample.c]: thanks for adding reference to the github issue for this change but also mention may be one or two lines to describe what problem are you solving here?

File: iothub_client/devdoc/requirement_docs/iothubclient_requirements.md [hunk 3]
```diff
@@ -378,12 +378,14 @@ Options handled by IoTHubClient_SetOption:
 
 **SRS_IOTHUBCLIENT_41_003: [** The value of `OPTION_DO_WORK_FREQUENCY_IN_MS` shall be limited to 100 to follow SDK best practices by not reducing the DoWork frequency below 10 Hz **]** 
 
-**SRS_IOTHUBCLIENT_41_004: [** If `currentMessageTimeout` is not less than `do_work_freq_ms`, `IotHubClientCore_SetOption` shall return `IOTHUB_CLIENT_ERROR` **]**
+**SRS_IOTHUBCLIENT_41_004: [** If `currentMessageTimeout` is less than `do_work_freq_ms`, `IotHubClientCore_SetOption` shall return `IOTHUB_CLIENT_ERROR` **]**
```
Comment (kiranpradeep) [inline iothub_client/devdoc/requirement_docs/iothubclient_requirements.md]: 1. Isn't [ `currentMessageTimeout` is not greater than `do_work_freq_ms`], the correct usage?
2. Isn't IOTHUB_CLIENT_INVALID_ARG, the more appropriate return value?

Comment (YoDaMa) [reply iothub_client/devdoc/requirement_docs/iothubclient_requirements.md]: I agree. Updating comments to reflect this.

**Comment by massand** [comment]
@YoDaMa can you point out the negative unit test where we are validating the dowork freq against the message timeout? The

File: iothub_client/src/iothub_client_core.c [hunk 4]
```diff
@@ -1734,11 +1735,10 @@ IOTHUB_CLIENT_RESULT IoTHubClientCore_SetOption(IOTHUB_CLIENT_CORE_HANDLE iotHub
             /* Codes_SRS_IOTHUBCLIENT_41_005: [ If parameter `optionName` is `OPTION_MESSAGE_TIMEOUT` then `IoTHubClientCore_SetOption` shall set `currentMessageTimeout` parameter of `IoTHubClientInstance` ]*/
             else if (strcmp(OPTION_MESSAGE_TIMEOUT, optionName) == 0)
             {
-                tickcounter_ms_t foo = (tickcounter_ms_t)(*(tickcounter_ms_t*)value);
-                iotHubClientInstance->currentMessageTimeout = foo;
+                iotHubClientInstance->currentMessageTimeout = (tickcounter_ms_t)(*(tickcounter_ms_t*)value);
```
Comment (massand) [inline iothub_client/src/iothub_client_core.c]: The extra (tickcounter_ms_t) cast is extra. You do the same on line 1720. 
Also, nit: please add a space in your pointer casts to be consistent. Currently there are few that are type* and some that are type *

Comment (YoDaMa) [reply iothub_client/src/iothub_client_core.c]: updated to reflect this

**Comment by YoDaMa** [comment]
@massand no negative unit test implemented yet. I'll add that. 

Also: OPTION_MESSAGE_TIMEOUT is a deprecated option. We don't advertise that anywhere except [[1]](https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/inc/iothub_client_options.h#L38). Should I also add a log warning on the iothub_client_core layer that this option is deprecated and only works with AMQP?

**Pull request merged**